web/packages/teleterm/src/ui/initUi.test.ts

```diff
@@ -0,0 +1,142 @@
+    ).not.toHaveBeenCalledWith(
+      expect.objectContaining({ kind: 'user-job-role' })
+    );
+  });
+
+  test('only `user-job-role` is shown when reporting was enabled earlier', async () => {
+    const mockedAppContext = new MockAppContext();
+    mockUsageReportingEnabled(mockedAppContext, { enabled: true });
+    mockOpenRegularDialog(mockedAppContext, dialog => {
+      if (dialog.kind === 'user-job-role') {
+        dialog.onSend('Engineer');
+      }
+    });
+
+    await initUi(mockedAppContext);
+
+    expect(
+      mockedAppContext.modalsService.openRegularDialog
+    ).not.toHaveBeenCalledWith(expect.objectContaining({ kind: 'usage-data' }));
+    expect(
+      mockedAppContext.modalsService.openRegularDialog
+    ).toHaveBeenCalledWith(expect.objectContaining({ kind: 'user-job-role' }));
+  });
+
+  test('no dialog is shown when reporting was disabled earlier', async () => {
+    const mockedAppContext = new MockAppContext();
+    mockUsageReportingEnabled(mockedAppContext, { enabled: false });
+    mockOpenRegularDialog(mockedAppContext);
+
+    await initUi(mockedAppContext);
+
+    expect(
+      mockedAppContext.modalsService.openRegularDialog
+    ).not.toHaveBeenCalledWith(expect.objectContaining({ kind: 'usage-data' }));
+    expect(
+      mockedAppContext.modalsService.openRegularDialog
+    ).not.toHaveBeenCalledWith(
+      expect.objectContaining({ kind: 'user-job-role' })
+    );
+  });
+
+  test('no dialog is shown when reporting was enabled and user was asked about job role earlier', async () => {
+    const mockedAppContext = new MockAppContext();
+    mockUsageReportingEnabled(mockedAppContext, { enabled: true });
+    jest
+      .spyOn(mockedAppContext.statePersistenceService, 'getUsageReportingState')
+      .mockImplementation(() => ({ askedForUserJobRole: true }));
+    mockOpenRegularDialog(mockedAppContext);
+
+    await initUi(mockedAppContext);
+
+    expect(
+      mockedAppContext.modalsService.openRegularDialog
+    ).not.toHaveBeenCalledWith(expect.objectContaining({ kind: 'usage-data' }));
+    expect(
+      mockedAppContext.modalsService.openRegularDialog
+    ).not.toHaveBeenCalledWith(
+      expect.objectContaining({ kind: 'user-job-role' })
+    );
+  });
+});
+
+function mockUsageReportingEnabled(
+  mockedAppContext: AppContext,
+  options: { enabled: boolean }
+) {
+  jest
+    .spyOn(mockedAppContext.mainProcessClient.configService, 'get')
+    .mockImplementation(key => {
+      if (key === 'usageReporting.enabled') {
+        return {
+          value: options.enabled,
+          metadata: {
+            isStored: true,
+          },
+        };
+      }
+    });
+}
+
+function mockOpenRegularDialog(
+  mockedAppContext: AppContext,
+  implementation?: (dialog: Dialog) => void
+) {
+  jest
+    .spyOn(mockedAppContext.modalsService, 'openRegularDialog')
+    .mockImplementation(dialog => {
+      implementation?.(dialog);
+      return {
+        closeDialog() {},
+      };
+    });
```

Comment:
If someone introduces a regression, for example they revert your change from `initUi` and call `setUpUsageReporting` first, the way the mocks are set up will make that `await setUpUsageReporting` hang until the timeout and return a less useful error message.

It's too bad that we don't have a uniform API for this between dialogs, but one thing we could do is this:

```suggestion
    .mockImplementation(dialog => {
      implementation?.(dialog);
      dialog['onCancel']?.();
      return {
        closeDialog() {},
      };
    });
```

This will immediately resolve any promise if it wasn't already resolved and let the test fail on an unmet expectation, rather than on exceeding the timeout.

Reply:
Good catch, done.